Initial draft of staking, rewards & penalties. (#36)
aeternity/hyperchains-whitepaper: staking.md

# Staking

#### **Constants:**

- **`MINIMUM_STAKE`**: The minimum amount of tokens required to be eligible for leader selection in any given epoch.

#### **Participant's Balances:**

- **Total Balance (TB)**: The total number of tokens a participant has deposited in the staking contract via the `stake()` function.
- **Locked Balance (LB)**: The max of staked amounts over the **current staking cycle**, i.e., epochs `N` to `N + C`, where `C` is the number of epochs in the staking cycle (in our case, `C = 3`).
- **Available Balance (AB)**: The portion of the Total Balance that is not locked in the current staking cycle.
  - **AB = TB - LB**
- **Staking Schedule (orderdlist ({epoch, amount}))**: A mapping of epochs to the amount of tokens the participant has committed to stake in each of those epochs and onwards if not changed.

Automatic Restaking: Unless adjusted, the stake for an epoch automatically continues into future cycles.

> 

Reviewer: Given the auto restaking and staking schedule as described: If I understood this correctly, the mapping would then be something like `map(address => map(epoch => int))` 

- If we would store that as a comprehensive data set on chain with all the information for every epoch, this would grow very, very fast and would have to be "maintained" by _somebody/something_ **each epoch**.
- If we would store this as an _incomplete_ data set, containing only the epochs the stake was staked (and GC when unstaked), we could technically calculate the full schedule, but the computational effort would grow over time for that, too.

Author: We only need to keep track of the staking within one cycle (4 epochs) at any specific height. We also only need to recalculate this when a call to adjusStake is made.

Author: Ill try to make this explicit in the document and show the algorithm.

---

#### **API Functions**

1. **`deposit()`** (Payable Endpoint)
   - **Description**: Allows a participant to deposit tokens into the staking contract, increasing their Total Balance (TB) and Available Balance (AB).
   - **Parameters**: None (the amount is determined by the value of tokens sent with the transaction).
   - **When Callable**: At any time.
   - **Behavior**:
     - Increases TB by the amount of tokens sent.
     - Increases AB by the same amount.

2. **`withdraw(amount)`**
   - **Description**: Allows a participant to withdraw tokens from their Available Balance (AB).
   - **Parameters**:
     - `amount`: The amount of tokens to withdraw.
   - **When Callable**: At any time.
   - **Constraints**:
     - `amount` must be less than or equal to AB.
     - Cannot withdraw tokens locked in the current staking cycle (i.e., tokens committed in SS for epochs `N` to `N + C`).

3. **`adjustStake(amount:integer)`**
   - **Description**: Adjusts the amount of tokens a participant wishes to stake for the **next staking cycle**.
   - **Parameters**:
     - `amount`: The amount to adjust (positive to increase stake, negative to decrease stake).

>    - **When Callable**: Only during the **Staking Epoch**.

Reviewer: I could use clarification here, isn't it always _some_ epoch that you could commit a stake in?

Author: Yes. Every epoch is a staking epoch for some staking cycle, so you can always do this, but it only affects the cycle for which you are in the staking epoch. Ill try to make it clearer.

   - **When Callable**: Only during the **Staking Epoch**.
   - **Constraints**:
     - Staking made in the current epoch `N` affect the staking cycle starting in epoch `N`, with leader election in epoch  `N + 1`.
     - **Positive `amount`**:
       - Must have sufficient AB to cover the increase.
     - **Minimum Stake Requirement**:
       - After adjustment, if the participant's stake is below `MINIMUM_STAKE`, they will not be eligible for leader selection in the upcoming cycle.
   - **Behavior**:
     - Updates the participant's stake commitment for the next staking cycle.
     - Recalculats LB:
       - **LB = amount**
     - Recalculates AB:
       - **AB = TB - LB**

4. **`getStakedAmount(epoch)`**
   - **Description**: Returns the amount of tokens the participant has committed to stake in the staking cycle starting at `epoch` and ending in `epoch + 3`
   - **Parameters**:
     - `epoch`: The epoch number.
   - **When Callable**: At any time.

5. **`getAvailableBalance()`**
   - **Description**: Returns the participant's current Available Balance (AB).
   - **Parameters**: None.
   - **When Callable**: At any time.

> 

Reviewer: Does the latter `alternative` intersect with the first one some how or are they to be viewed disjunctively? If latter case, it is missing some form of a preceding deposit function, among other things. Or is this alternative just an alternative for the previous `adjustStake` function , and the numbering is just a bit messed up? In that case, we would miss some way to reduce the locked stake, I guess ?

Author: We only need one of the functions, 'adjustStake' or 'stake.' 
Either it is relative to the current stake or absolute. In both cases you need to use deposit and withdraw to ensure you have the right amount of tokens in the staking contract.

Reviewer 2: That reminds me of a long and sometimes confusing discussion with @thepiwo regarding how you could alter allowances in the AEX-9 contract. There were some interesting subtleties there, not sure if they are here also.

  - Epoch 13: Payout Epoch
- **Participant**: Alice
- **`MINIMUM_STAKE`**: 100 tokens

#### **Alice's Balances**

- **Total Balance (TB)**: 1000 tokens (Alice has already deposited these via `deposit()`).
- **Available Balance (AB)**: 1000 tokens (initially, no tokens are locked).
- **Locked Balance (LB)**: 0 tokens.

#### **Epoch 10 (Staking Epoch)**

- **Action 1**: Alice decides to stake 400 tokens for the next staking cycle (epochs 11 to 13).
  - She calls `adjustStake(400)`.
  - **AB** decreases by 400 tokens:
    - AB = 1000 - 400 = 600 tokens.
  - **Stake Commitment**:
    - Alice's stake commitment for the next staking cycle is now 400 tokens.

- **Action 2**: Alice withdraws 100 tokens.
  - She calls `withdraw(100)`.
  - **TB** decreases by 100 tokens:
    - TB = 1000 - 100 = 900 tokens.
  - **AB** decreases by 100 tokens:
    - AB = 600 - 100 = 500 tokens.

- **Locked Balance (LB)**:
  - Set at 400 from the call to adjustStake till the end of the Staking Epoch.

#### **Epoch 11 (Leader Election Epoch)**

- **Transition to New Staking Cycle**:
  - Alice's stake of 400 tokens becomes locked for the duration of the staking cycle (epochs 11 to 13).

>   - **LB** increases by 400 tokens:

Reviewer: Coming from the question to the staking schedule: Is something imperatively setting this value, or is it  calculated by whatever entity that needs it? This would mean we would have to iterate over all the stakings this user ever did to find the latest, relevant ones - or do we have efficient ways to access only the 'last couple' elements of a list?

    - LB = 0 + 400 = 400 tokens.
  - **AB** remains at 500 tokens:
    - AB = TB - LB = 900 - 400 = 500 tokens.

- **Adjusting Stake for Next Cycle**:
  - Alice can now adjust her stake for the **next staking cycle** (epochs 11 to 14) during the Staking Epoch of that cycle.
 - She calls `adjustStake(-400)` this will not free up her available balance right away but in epoch 14.

#### **Epochs 11 to 13**

- **Funds Locked**:
  - The 400 tokens are locked and cannot be withdrawn until the end of the Payout Epoch (epoch 13).
  - **AB** remains at 500 tokens.

- **Actions**:
  - Alice cannot adjust her current stake but can plan for the next cycle.

#### **Epoch 14 (Next Staking Epoch)**

- **Unlocking Funds**:

>   - At the end of epoch 13 (Payout Epoch), the 400 tokens become unlocked.

Reviewer: If the 400 tokens _become unlocked_ somehow, this means that they explicitly would need to be restaked, which would be contradictory to the `Automatic Re-Staking ` ? Or do they become _unlockable_, by being 'in the next' staking epoch now (E + 4), but still being actually _staked_ one way or another ? Then the definition of `locked balance` and `available balance` would have to be ammended though, as it wouldn't always correspond to something like 'staked balance' `SB` : You could have committed 400 in E11,  `LB` = 400 == ``SB` = 400 in E13, but theoretically in E14 have `SB` = 400 and `LB = 0`, because those 400 are not locked anymore as you can withdraw them now due to the Epoch being E + 4.

If it is really the case as written though, then we technically cannot speak of automated re-staking I guess ?

Author: Alice unstaked the 400 in epoch 11 so in epoch 14 they become unlocked.
Ill try to clarify.

# Staking

#### **Constants:**

- **`MINIMUM_STAKE`**: The minimum amount of tokens required to be eligible for leader selection in any given epoch.

#### **Participant's Balances:**

- **Total Balance (TB)**: The total number of tokens a participant has deposited in the staking contract via the `stake()` function.

> - **Locked Balance (LB)**: The max of staked amounts over the **current staking cycle**, i.e., epochs `N` to `N + C`, where `C` is the number of epochs in the staking cycle (in our case, `C = 3`).

Reviewer 3: it's not discussed in detail here in the document, but don't we want to have an additional locking period before funds can be withdrawn?

Reviewer 4: The rewards are paid in the epoch after the block production and after voting for a next block. My assumption is that the voting is the key here in not having to wait any longer to pay out?

Author: This is the minimum time needed for locking. It is discussed under locking that longer locking could be configured.

## Locking

In the current proposal a stake is only locked for one cycle.
We could consider adding a network/hyperchain parameter stating the number of epochs or cycles that
a stake is locked.
In this case the `adjustStake` function would need to be updated to keep track of the locking for
the whole locking period.

- **Available Balance (AB)**: The portion of the Total Balance that is not locked in the current staking cycle.
  - **AB = TB - LB**

> - **Staking Schedule (SC)**: A mapping of epochs to the amount of tokens the participant has committed to stake in each of those epochs and onwards if not changed. As an orderd list ({epoch, stake amount, locked amount})

Reviewer 3: Is there a specific reason to have it as orderd list over map epoch => { stake amount, locked amount } or similar

Author: A list is easier to GC, but that is an implementation detail. Here, I am describing the functionality; an order list or map doesn't matter. I just used a list to wrap my head around what is needed,

Automatic Restaking: Unless adjusted, the stake for an epoch automatically continues into future cycles.

---

#### **API Functions**

1. **`deposit(producer:pubkey)`** (Payable Endpoint)

>    - **Description**: Allows a participant to deposit tokens into the staking contract, increasing their Total Balance (TB) and Available Balance (AB).

Reviewer 3: > increasing their Total Balance (TB)

is above defined as "deposited in the staking contract via the `stake()` function", but should probably be explained either there or here that this also includes the automatic restaking balance for which the `stake()` function hasn't been called explicitly

   - **Parameters**:
     - producer - the pubkey of the producer who can become leader and then produce and sign blocks.
     -  (the amount is determined by the value of tokens sent with the transaction).
   - **When Callable**: At any time.
   - **Behavior**:
     - Increases TB by the amount of tokens sent.
     - Increases AB by the same amount.

2. **`withdraw(producer:pubkey, amount:unsigned integer)`**
   - **Description**: Allows a participant to withdraw tokens from their Available Balance (AB).
   - **Parameters**:

>      - `producer`: The pubkey of the producer, must match the signature of the call transaction (do we need it).

Reviewer 2: So producer is `Call.caller` then? Let's try to avoid manual juggling of signatures.

Reviewer 3: agree, `Call.caller` can withdraw, if there is AB for him, no manual signature or parameter necessary

     - `amount`: The amount of tokens to withdraw.
   - **When Callable**: At any time.
   - **Constraints**:
     - `amount` must be less than or equal to AB.
     - Cannot withdraw tokens locked in the current staking cycle (i.e., tokens committed in SS for epochs `N` to `N + C`).

3. **`adjustStake(producer:pubkey, amount:integer)`**
   - **Description**: Adjusts the amount of tokens a participant wishes to stake for the **next staking cycle**.
   - **Parameters**:

>      - `producer`: The pubkey of the producer, must match the signature of the call transaction (do we need it).

Reviewer 3: same as above https://github.com/aeternity/hyperchains-whitepaper/pull/36/files#r1855533592

     - `amount`: The amount to adjust (positive to increase stake, negative to decrease stake).

>    - **When Callable**: Only during the **Staking Epoch** of a cycle, but every epoch is the staking epoch of some cycle, so syou can always call it.

Reviewer 2: ```suggestion
   - **When Callable**: Only during the **Staking Epoch** of a cycle, but every epoch is the staking epoch of some cycle, so you can always call it.
```

   - **Constraints**:

>      - Adjustments made in the current epoch `N` affect the staking cycle starting in epoch `N`, with leader election in epoch  `N + 1`.

Reviewer 2: I'm not sure what `N+1` means here, if you alter your stake that will earliest affect the leader schedule three or four epochs down the line?

Author: Fixed I think.

   - Adjusts the participant's stake commitment for the next staking cycle.
   - **Positive Adjustment:**
     - Decreases AB by `amount`.
     - Increases the stake commitment for the next staking cycle by `amount`.
   - **Negative Adjustment:**
     - Increases AB by `-amount`.
     - Decreases the stake commitment for the next staking cycle by `-amount`.
   - **AB Recalculation:**
     - **AB = TB - LB**

3. **Withdrawal (`withdraw()`):**
   - Decreases TB and AB by `amount`.
   - Cannot affect tokens locked in the any of the 4 concurrent staking cycles.

4. **Epoch Transition:**
   - At the end of the Payout Epoch (`N + 3`):
     - Tokens locked in the current staking cycle become available.
     - **LB** decreases by the amount of tokens that were locked.
     - **AB** increases accordingly.

---

### **Updated Example Scenario**

Let's illustrate how the balances change over epochs with this updated model.

#### **Constants and Parameters**

- **Current Epoch (`N`)**: 10 (Staking Epoch)
- **Staking Cycle Length (`C`)**: 4 epochs (epochs 10 to 13)
  - Epoch 10: Staking Epoch
  - Epoch 11: Leader Election Epoch
  - Epoch 12: Block Production Epoch

>   - Epoch 13: Payout Epoch

Reviewer 2: I think it would simplify the discussion if this aligns with the other document and the implementation...

- **Insufficient AB for Adjusting Stake:**
  - If a participant tries to increase their stake but does not have sufficient AB, they must first increase TB via `deposit()`.

- **Withdrawal Limitations:**
  - Participants cannot withdraw tokens that are locked in the current staking cycle.
  - Attempting to withdraw more than AB will result in an error.

- **Automatic Re-Staking:**
  - Tokens do not automatically unlock, instead if no change in staking is registerd they remain at stake.
  - Participants must call `adjustStake()` during the next Staking Epoch to free tokens during the next staking cycle.

---

### **Example with Varying Stakes Across Epochs**

If participants can adjust their stake for each epoch within the staking cycle, the LB calculation becomes the maximum over the epochs.

Suppose Alice stake different amounts in each epoch:

- **Epoch 11**: 500 tokens
- **Epoch 12**: 300 tokens
- **Epoch 13**: 400 tokens

Then in epoch 13, we are still in the cycle started in epoch 11.

- **LB = max(SS[11], SS[12], SS[13])**
- **LB = max(500,300,400) = 500 tokens**

## Locking

In the current proposal a stake is only locked for one cycle.

> We could consider adding a network/hyperchain parameter stating the numner of epochs or cycles that

Reviewer 2: ```suggestion
We could consider adding a network/hyperchain parameter stating the number of epochs or cycles that
```

- **Missed Blocks**: If a validator fails to produce a block in their assigned time slot, this results in a **missing block**.

- **Producing a 'Hole'**:
  - The subsequent validator in the leader schedule is allowed to produce a special type of block called a **'hole'**.
  - **Hole Characteristics**:
    - Represents the absence of the expected block.
    - Contains no transactions.
    - Serves as a placeholder to maintain the continuity of the blockchain.
  - **No Rewards for Holes**:
    - Validators producing holes do **not** receive any rewards for these blocks.
    - The missed block's potential rewards are effectively forfeited, including the award of fees for following the previous block.

## Reward Distribution for Correct Blocks

When a validator successfully produces a block on time, the rewards are distributed as follows:

### Components of Rewards

1. **Block Fees**: Transaction fees collected from transactions included in the block.
2. **Epoch Coinbase**: Newly minted tokens allocated for that epoch.

### Distribution Breakdown

- **Block Producer (Current Validator)**:

>   - Receives **X% of the block fees**, X defaults to 75% but is configurable in genisis.

Reviewer 2: ```suggestion
  - Receives **X% of the block fees**, X defaults to 75% but is configurable in genesis.
```

    - Changes to the reward parameters can be proposed and voted upon by stakeholders.
    - Ensures that adjustments reflect the consensus of the network participants.

# Penalties & Slashable Offenses

Here we describe the effects of penalties on staking and rewards for validators who commit slashable offenses. It outlines how these penalties impact a validator's staked tokens, their eligibility for rewards, and their future participation in the network.

## Overview of Penalties and Slashable Events

Penalties are enforced to deter malicious actions or protocol violations. **Slashable events** are actions that result in the forfeiture of a validator's stake, reputation, or other penalties to maintain network integrity and fairness. Any participant can submit proof of such wrongdoing, ensuring a decentralized and fair enforcement mechanism.

### Slashable Offenses

1. **Producing Two Versions of a Block at a Specific Height (Double-Spending Attack)**
2. **Double Voting**
3. **Ignoring Votes**
4. **Consistently Failing to Submit Blocks on Time**

### Minor Offenses
1. **Ignoring the `finalize_epoch` Fork**
2. **Ignoring a Correctly Pinned Fork**

>  For minor offenses like ignoring forks (items 5 and 6), the protocol ignore the validator's actions without imposing penalties.

Reviewer 2: ```suggestion
 For minor offenses like ignoring forks (items 1 and 2), the protocol ignore the validator's actions without imposing penalties.
```

### Details TBD

> The exact detials of what these offenses and how proof of these offense should look and the penalties is yet to be decided.

Reviewer 2: ```suggestion
The exact details of what these offenses and how proof of these offense should look and the penalties is yet to be decided.
```

- **Available Balance (AB)**: Calculated as the difference between TB and LB.
  - **AB = TB - LB**

### **Lifecycle of Tokens in the Staking Contract**

1. **Staking Tokens (`deposit()`):**
   - Increases TB and AB.
   - No immediate effect on staking commitments until `adjustStake()` (or `stake()`) is called.

2. **Adjusting Stake (`adjustStake()`):**
   - Adjusts the participant's stake commitment for the next staking cycle.
   - **Positive Adjustment:**
     - Decreases AB by `amount`.
     - Increases the stake commitment for the next staking cycle by `amount`.
   - **Negative Adjustment:**
     - Increases AB by `-amount`.
     - Decreases the stake commitment for the next staking cycle by `-amount`.
   - **AB Recalculation:**
     - **AB = TB - LB**

3. **Withdrawal (`withdraw()`):**
   - Decreases TB and AB by `amount`.
   - Cannot affect tokens locked in the any of the 5 concurrent staking cycles.

4. **Epoch Transition:**
   - At the end of the Payout Epoch (`N + 4`):
     - Tokens locked in the current staking cycle become available.
     - **LB** decreases by the amount of tokens that were locked.

>      - **AB** increases accordingly.

Reviewer 2: All of this guarded by some max/min-logic w.r.t. the other "active" epochs, right?

Author: Yes there is a note later:

### **Example with Varying Stakes Across Epochs**

If participants can adjust their stake for each epoch within the staking cycle, the LB calculation becomes the maximum over the epochs.

Suppose Alice stake different amounts in each epoch:

- **Epoch 11**: 500 tokens
- **Epoch 12**: 300 tokens
- **Epoch 13**: 400 tokens

Then in epoch 13, we are still in the cycle started in epoch 11.

- **LB = max(SS[11], SS[12], SS[13])**
- **LB = max(500,300,400) = 500 tokens**

And the description of  the adjustStake functionality.

- **Alice's Balances**:
  - No changes occur in this epoch.
  - **AB** remains at **500 tokens**.
  - **LB** remains at **400 tokens**.

---

### **Epoch 12 (Leader Election Epoch)**

- Alice's stake of **400 tokens** stays **locked** for the duration of the staking cycle (epochs **12** to **14**).
- **Alice's Balances**:
  - **LB** = **400 tokens**.
  - **AB** = **500 tokens**.

#### **Adjusting Stake for Next Cycle**
- Alice can now plan her stake for the next next staking cycle (block production in epoch 15) and decides on no change.
- She decides to **withdraw** her stake after the current cycle:
  - She calls `adjustStake(-400)`.
  - This action schedules her stake to be reduced by 400 tokens after the current cycle ends.
- **Note**: The **AB** and **LB** remain unchanged until the stake is unlocked.

---

### **Epoch 13 (Block Production Epoch)**

- **Block Production**:
  - Alice participates in block production based on her **locked stake** of **400 tokens**.
- **Alice's Balances**:
  - No changes occur in this epoch.
  - **AB** remains at **500 tokens**.
  - **LB** remains at **400 tokens**.

---

### **Epoch 14 (Payout Epoch)**

#### **Rewards Distribution**

> - Alice receives rewards based on her participation, paid out directly to her account.

Reviewer 2: I think rewards will be payed out to a/the contract or else the delegated staking will not work?

Author: Yes, if you are a "delegated" staker your delegation contract should be the one calling the deposit and stake functions and payments of rewards should go there.

Author: I added some info on Alice.

- **System Process**:
  - The system waits for the parent chain's hash for entropy.
- **Alice's Balances**:
  - The stake become locked
  - **AB** remains at **400 tokens**.
  - **LB** is set to **500 tokens**.

---

### **Epoch 17 (Leader Election Epoch)**

- **Alice's Balances**:
  - No changes occur.
  - **AB** remains at **400 tokens**.
  - **LB** is set to **500 tokens**.

---

### **Epochs 17 to 19**

- **Block Production and Payout**:
  - Alice participates in block production.
  - Rewards are earned and distributed in the Payout Epoch (epoch **19**).
- **Actions**:
  - Alice cannot adjust her **current** stake but can plan for the **next** cycle during Epoch 20.

---
Certainly! Based on your updated scenario where the stake becomes locked directly after the end of the **Staking Epoch** (at the start of the **Entropy Epoch**), here is the updated table reflecting Alice's balances over the epochs:

---

### **Summary of Alice's Balances Over Epochs**

| Epoch | TB (tokens) | AB (tokens) | LB (tokens) | Notes                                                 |
|-------|-------------|-------------|-------------|-------------------------------------------------------|

> | 09    | 1000        | 500         |   0         | Inital balance 1000 tokens                            |

Reviewer 2: ```suggestion
| 09    | 1000        | 500         |   0         | Initial balance 1000 tokens                           |
```

| 15    | 900  | 0    | 400  |  500 | `{15, 500, 0}` , `{16, 500, 500}`, `{17, 500, 500}`, `{18, 500, 500}`, `{19, 500, 500}`  | New Stake of 500                                    |
| 16    | 900  | 500  | 400  |  500 | `{15, 500, 0}` , `{16, 500, 500}`, `{17, 500, 500}`, `{18, 500, 500}`, `{19, 500, 500}`  |                                                     |
---

### **Edge Cases and Additional Notes**

- **Adjusting Stake During Staking Epoch:**
  - Participants can only adjust their stake for the **current staking cycle** during the Staking Epoch.
  - Once the Staking Epoch ends, the stake commitments are locked till the end of the cycle.

- **Insufficient AB for Adjusting Stake:**
  - If a participant tries to increase their stake but does not have sufficient AB, they must first increase TB via `deposit()`.

- **Withdrawal Limitations:**
  - Participants cannot withdraw tokens that are locked in the current staking cycle.
  - Attempting to withdraw more than AB will result in an error.

- **Automatic Re-Staking:**
  - Tokens do not automatically unlock, instead if no change in staking is registerd they remain at stake.

>   - Participants must call `adjustStake()` during the next Staking Epoch to free tokens during the next staking cycle.

Reviewer 2: This might not make it into the first version.

Author: No, probably lots of things not making it into the first version.
